### PR TITLE
`Development`: Reduce client main bundle size and introduce caching

### PR DESCRIPTION
**Pull request by ge65cer**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.

### Motivation and Context

The current main bundle size is 539 KB (gzipped). In general, we should try to reduce the bundle size wherever possible to speed up the application load time.

### Description

This PR works towards reducing the main bundle size and speeding up the build process ([now 532 KB](https://bamboo.ase.in.tum.de/browse/ARTEMIS-WEBAPP4169-11/artifact/shared/webpack-bundle-report/stats.html)).
The dependency `papaparse` is now lazy loaded and was removed from the main bundle (required a workaround because of some weird behaviour when importing it in an abstract superclass).
It also optimizes the imports from `ngx-charts` and `crypto-js` by limiting it to the modules that are actually used.
Ideally, using tree-shaking the Angular Ivy engine should now be able to throw out any modules not used in Artemis.
This [does not seem to work](https://github.com/swimlane/ngx-charts/issues/1765) for `ngx-charts`, yet.

Most importantly, this PR introduces caching for all `*.js, *.css` files (as well as for `/i18n`).
The caching period can be configured using `jhipster.http.cache.timeToLiveInDays`.
Due to [an issue in JHipster](https://github.com/jhipster/generator-jhipster/issues/19102), the server currently does not set the caching period correctly (milliseconds instead of sections, factor 1000).
Because all JS and CSS files are delivered with a hash in the filename, this shouldn't be a big issue and cache busting should work automatically.

Thanks to @akesfeden, @willergeorg and @bassner for their discussion/inputs on this PR.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate through different components using charts
3. Ensure that all charts are still rendered without issues
4. Ensure all components related to the “grading key” work correctly
5. Ensure that the resource files (`*.js, *.css, /i18n`) are delivered with a `Cache-Control` header
6. Deploy a different PR to the test server to ensure that cache busting works (you should _not_ have to force-reload, usually Shift+F5, to receive the updated resource files)

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

File: .browserslistrc [hunk 1]
```diff
@@ -10,9 +10,11 @@
 
 last 1 Chrome version
 last 1 Firefox version
-last 2 Edge major versions
-last 2 Safari major versions
+last 1 Edge major versions
+last 1 Safari major versions
 last 2 iOS major versions
```
Comment (JohannesStoehr) [inline .browserslistrc:15]: Doesn't apple release a new version of Safari and iOS every year?
Why do we support iOS longer than Safari?

File: src/main/webapp/app/grading-system/base-grading-system/base-grading-system.component.ts [hunk 2]
```diff
@@ -695,20 +695,12 @@ export abstract class BaseGradingSystemComponent implements OnInit {
 
     /**
      * Parse CSV file to a list of CsvGradeStep objects
+     * This method uses the 'papaparse' dependency, but Angular can not lazy load it in an abstract class.
+     * Therefore, the implementation is moved to the concrete classes (code duplication for the benefit of reducing main bundle size).
      * @param csvFile the read csv file
-     * @private
+     * @abstract
```
Comment (JohannesStoehr) [inline src/main/webapp/app/grading-system/base-grading-system/base-grading-system.component.ts:701]: Why did you duplicate the methods in the subclasses?

**Review by JohannesStoehr — COMMENTED** [review]
Left two questions:

File: src/main/webapp/app/grading-system/base-grading-system/base-grading-system.component.ts (hunk 2 above)
Comment (ge65cer) [reply src/main/webapp/app/grading-system/base-grading-system/base-grading-system.component.ts:701]: > This method uses the 'papaparse' dependency, but Angular can not lazy load it in an abstract class. Therefore, the implementation is moved to the concrete classes (code duplication for the benefit of reducing main bundle size).

File: .browserslistrc (hunk 1 above)
Comment (ge65cer) [reply .browserslistrc:15]: I went by the general rule “desktop 1 last version, mobile 2 last versions”.

**Review by willergeorg — APPROVED** [review]
Tested during testing session. Did not percieve any distorted layout. Charts look good as well. 👍

**Review by Dominik-Weinzierl — APPROVED** [review]
Looks good - besides the discussed points :+1:

**Review by mallgaier — DISMISSED** [review]
Tested during testing session

**Pull request merged**